### PR TITLE
fix for python 3.10 for gradient opinfo

### LINKED ISSUE #67612: [testing] `gradient` OpInfo tests fail in Python 3.10
EDIT: The failure is due to update in Python 3.10 behaviour.

The following OpInfo tests fail locally (with Python 3.10) but pass on CI for `gradient` op

https://github.com/pytorch/pytorch/blob/97f29bda59deab8c063cf01f0a8ff4321b93c55e/torch/testing/_internal/common_methods_invocations.py#L8277-L8282

<details>

<summary>Local failure log</summary>

```
_____________________________________________________________________________________ TestCommonCPU.test_dtypes_gradient_cpu _____________________________________________________________________________________
Traceback (most recent call last):
  File "/home/kshiteej/Pytorch/pytorch_testing_native_device/test/test_ops.py", line 158, in test_dtypes
    self.assertEqual(supported_dtypes, claimed_supported, msg=msg)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 1946, in assertEqual
    super().assertEqual(x, y, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/unittest/case.py", line 837, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/unittest/case.py", line 1097, in assertSetEqual
    self.fail(self._formatMessage(msg, standardMsg))
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/unittest/case.py", line 667, in fail
    raise self.failureException(msg)
AssertionError: Items in the second set but not the first:
torch.int32
torch.int16
torch.int8
torch.int64 : Attempted to compare [set] types: Expected: {torch.float64, torch.complex64, torch.bfloat16, torch.complex128, torch.float16, torch.float32}; Actual: {torch.float64, torch.int16, torch.int32, torch.complex64, torch.int64, torch.bfloat16, torch.complex128, torch.float16, torch.float32, torch.int8}.
The supported dtypes for gradient on cpu according to its OpInfo are
        {torch.float64, torch.int16, torch.int32, torch.complex64, torch.int64, torch.bfloat16, torch.complex128, torch.float16, torch.float32, torch.int8}, but the detected supported dtypes are {torch.float64, torch.complex64, torch.bfloat16, torch.complex128, torch.float16, torch.float32}.
        The following dtypes should be removed from the OpInfo: {torch.int32, torch.int16, torch.int8, torch.int64}.
__________________________________________________________________________ TestCommonCPU.test_noncontiguous_samples_gradient_cpu_int64 ___________________________________________________________________________
Traceback (most recent call last):
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_methods_invocations.py", line 747, in sample_inputs
    samples = self.sample_inputs_func(self, device, dtype, requires_grad, **kwargs)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_methods_invocations.py", line 2683, in sample_inputs_gradient
    a = torch.tensor(coords, dtype=dtype, device=device)
TypeError: 'float' object cannot be interpreted as an integer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kshiteej/Pytorch/pytorch_testing_native_device/test/test_ops.py", line 220, in test_noncontiguous_samples
    sample_inputs = op.sample_inputs(device, dtype, requires_grad=test_grad)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_methods_invocations.py", line 749, in sample_inputs
    samples = self.sample_inputs_func(self, device, dtype, requires_grad)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_methods_invocations.py", line 2683, in sample_inputs_gradient
    a = torch.tensor(coords, dtype=dtype, device=device)
TypeError: 'float' object cannot be interpreted as an integer
_________________________________________________________________________ TestCommonMETA.test_noncontiguous_samples_gradient_meta_int64 __________________________________________________________________________
Traceback (most recent call last):
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_methods_invocations.py", line 747, in sample_inputs
    samples = self.sample_inputs_func(self, device, dtype, requires_grad, **kwargs)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_methods_invocations.py", line 2683, in sample_inputs_gradient
    a = torch.tensor(coords, dtype=dtype, device=device)
TypeError: 'float' object cannot be interpreted as an integer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kshiteej/Pytorch/pytorch_testing_native_device/test/test_ops.py", line 220, in test_noncontiguous_samples
    sample_inputs = op.sample_inputs(device, dtype, requires_grad=test_grad)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_methods_invocations.py", line 749, in sample_inputs
    samples = self.sample_inputs_func(self, device, dtype, requires_grad)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_methods_invocations.py", line 2683, in sample_inputs_gradient
    a = torch.tensor(coords, dtype=dtype, device=device)
TypeError: 'float' object cannot be interpreted as an integer
____________________________________________________________________________________ TestCommonCUDA.test_dtypes_gradient_cuda ____________________________________________________________________________________
Traceback (most recent call last):
  File "/home/kshiteej/Pytorch/pytorch_testing_native_device/test/test_ops.py", line 158, in test_dtypes
    self.assertEqual(supported_dtypes, claimed_supported, msg=msg)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 1946, in assertEqual
    super().assertEqual(x, y, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/unittest/case.py", line 837, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/unittest/case.py", line 1097, in assertSetEqual
    self.fail(self._formatMessage(msg, standardMsg))
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/unittest/case.py", line 667, in fail
    raise self.failureException(msg)
AssertionError: Items in the second set but not the first:
torch.int32
torch.int16
torch.int8
torch.int64 : Attempted to compare [set] types: Expected: {torch.float64, torch.complex64, torch.bfloat16, torch.complex128, torch.float16, torch.float32}; Actual: {torch.float64, torch.int16, torch.int32, torch.complex64, torch.int64, torch.bfloat16, torch.complex128, torch.float16, torch.float32, torch.int8}.
The supported dtypes for gradient on cuda according to its OpInfo are
        {torch.float64, torch.int16, torch.int32, torch.complex64, torch.int64, torch.bfloat16, torch.complex128, torch.float16, torch.float32, torch.int8}, but the detected supported dtypes are {torch.float64, torch.complex64, torch.bfloat16, torch.complex128, torch.float16, torch.float32}.
        The following dtypes should be removed from the OpInfo: {torch.int32, torch.int16, torch.int8, torch.int64}.
____________________________________________________________________________ TestCommonCUDA.test_multiple_devices_gradient_cuda_int64 ____________________________________________________________________________
Traceback (most recent call last):
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_methods_invocations.py", line 747, in sample_inputs
    samples = self.sample_inputs_func(self, device, dtype, requires_grad, **kwargs)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_methods_invocations.py", line 2683, in sample_inputs_gradient
    a = torch.tensor(coords, dtype=dtype, device=device)
TypeError: 'float' object cannot be interpreted as an integer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kshiteej/Pytorch/pytorch_testing_native_device/test/test_ops.py", line 188, in test_multiple_devices
    samples = op.sample_inputs(cuda_device, dtype)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_methods_invocations.py", line 749, in sample_inputs
    samples = self.sample_inputs_func(self, device, dtype, requires_grad)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_methods_invocations.py", line 2683, in sample_inputs_gradient
    a = torch.tensor(coords, dtype=dtype, device=device)
TypeError: 'float' object cannot be interpreted as an integer
_________________________________________________________________________ TestCommonCUDA.test_noncontiguous_samples_gradient_cuda_int64 __________________________________________________________________________
Traceback (most recent call last):
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_methods_invocations.py", line 747, in sample_inputs
    samples = self.sample_inputs_func(self, device, dtype, requires_grad, **kwargs)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_methods_invocations.py", line 2683, in sample_inputs_gradient
    a = torch.tensor(coords, dtype=dtype, device=device)
TypeError: 'float' object cannot be interpreted as an integer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kshiteej/Pytorch/pytorch_testing_native_device/test/test_ops.py", line 220, in test_noncontiguous_samples
    sample_inputs = op.sample_inputs(device, dtype, requires_grad=test_grad)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_methods_invocations.py", line 749, in sample_inputs
    samples = self.sample_inputs_func(self, device, dtype, requires_grad)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/testing/_internal/common_methods_invocations.py", line 2683, in sample_inputs_gradient
    a = torch.tensor(coords, dtype=dtype, device=device)
TypeError: 'float' object cannot be interpreted as an integer
==================================================================================
```

</details>

But looks like it works on CI

All tests on CI pass on this PR: https://github.com/pytorch/pytorch/runs/4058942158?check_suite_focus=true#step:16:1931

Cause of failure on local:
https://github.com/pytorch/pytorch/blob/97f29bda59deab8c063cf01f0a8ff4321b93c55e/torch/testing/_internal/common_methods_invocations.py#L2675-L2683

`a = torch.tensor(coords, dtype=dtype, device=device) ` is where error occurs. Here `dtype` could be int but `coords` are always float and hence the error.

```python
>>> t = torch.tensor([1.0], dtype=torch.int64)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'float' object cannot be interpreted as an integer
```

@mruberry any idea?

cc @mruberry

### PR DESCRIPTION
**Pull request by ankitaS11**
This PR fixes https://github.com/pytorch/pytorch/issues/67612 by creating a tensor first and then converting the dtype explicitly using `.to(dtype)` call.

Looking forward to your feedback and suggestions on this.

cc: @kshitij12345 @mruberry

### CODE REVIEW

**Comment by mruberry** [comment]
Thanks @ankitaS11!

**Comment by mruberry** [comment]
@saketh-are -- would you take a look at the underlying issue and review this PR? 

Are there other changes we should look at making?

**Review by saketh-are — APPROVED** [review]
@ankitaS11 thanks for writing up this fix! This looks great overall. I left a couple of minor comments inline for your consideration.

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -2709,8 +2709,11 @@ def sample_inputs_gradient(op_info, device, dtype, requires_grad):
         t = make_tensor(size, device, dtype, low=None, high=None, requires_grad=requires_grad)
         coordinates_tensor_list = []
         for coords in coordinates:
-            a = torch.tensor(coords, dtype=dtype, device=device)
-            coordinates_tensor_list.append(a)
+            # `coords` will always contain floating point values and Python 3.10 does not support
+            # implicit conversion to an integer using `__int__`
+            # See https://github.com/pytorch/pytorch/issues/67612 for more context
+            a = torch.tensor(coords, dtype=torch.float32, device=device)
```
[the remark this answers is not in the document]
Comment (ankitaS11) [reply torch/testing/_internal/common_methods_invocations.py]: Thank you @saketh-are. I have made the required changes in the recent commit.

**Review by saketh-are — APPROVED** [review]
Thanks @ankitaS11, this looks good to me!

**Pull request closed without merging**